keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,425 @@
+
+The scheduler was chosen as the primary owner of the mapping of pods to requested resources and because it already manages an internal view of pod resources that matches the external idealized view of pod resources. We also expect that the scheduler will have configuration (gates, tuning, customizations) that would have practical effects on the represenation of the resource model and that aligning these metrics with the scheduler will minimize configuration mismatches or vendor drift when things like pod overhead, mutable resource requests, forked schedulers, or other points of deviation arise. While other components could certainly host these metrics, it would be as or more expensive (due to the duplication of the pod cache and maintenance of the vendoring logic for shared resource model code) and would not offer a meaningful benefit to an administrator of Kubernetes. The metrics described in this proposal apply to the idealized view of pods that transition from pending (unscheduled) to running (scheduled) to completed (terminal), all of which are critical to the functioning of the scheduler. The metrics require no integration into the internal state of the scheduler or impose any additional requirements on the scheduler except for access to the pod informer - if the scheduler chooses to have a more efficient internal representation of the pod cache in the future the same data could be shared with the metrics reporting since both scheduler and the metrics will need an accurate representation of resources (even if many other fields are not needed).
+
+The metrics will be implemented via the collector pattern - when a scrape is requested the pod cache will be traversed and each metric will be calculated and reported in a streaming fashion, without caching or having to acquire locks. This is the standard pattern within Kubernetes for reporting high cardinality metrics and generally allows us to discount the cost of scrapes (estimate: one scrape every 15s might add 0.1 core of use for 50k pods, and we can control scape interval and reduce cost). The approach will add no additional memory usage overhead except for the transient allocations during iteration of the pod cache. Note that in other components we use this pattern to gather one or two orders of magnitude more metrics proportional to pods, so from a whole system perspective the impact is negligible and should have no impact on the scheduler performance.
```

Comment:
Just making sure I understood this correctly:
There is not going to be routine in the scheduler continuously running to export the metrics. They will only be collected when an external agents queries the endpoint.
I was not familiar with this approach, but I like it. It also saves us from having to offer a knob to enable/disable these metrics.

Reply:
Yes, think of this as the "advanced / high scale" model.  It plays extremely well to any component exposing metrics based on data in an existing cache.

1. Agent scrapes /metrics/foo
2. Endpoint handler invokes the collector
3. Collector iterates over a cache / other in memory store (although some implementations might make a call to disk or another component to get data, like the Kubelet does for cgroup data on the filesystem) and reports one metric per entry